feat: providers and models configuration

denishacquin opened the pull request:
Fixes #46 
Fixes #415 

# What

- Added ability to create / update providers from the frontend
  - Added provider list in settings
  - Client-side validation of configs
  - Client-side validation of unique constraints
  - Handling of misconfiguration cases in backend and UI
- Reworked the navigation to improve structure
  - New dedicated Validators tab
  - Profile becomes Settings
- Fixed a bug with a broken condition
- Fixed: selecting an account now closes the dropdown
- Fixed: integer validation on stake field that was breaking e2e test
- Moved account creation/delete directly in the account dropdown for simplification
- Fixed a scrolling issue on the home tab that was sometimes causing e2e tests to fail

Modals:
<img width="658" alt="image" src="https://github.com/user-attachments/assets/a8527096-5d93-4a88-a49d-fef057e5eec9">
<img width="737" alt="image" src="https://github.com/user-attachments/assets/650a4687-0203-4065-97be-6d23727e6004">
Nav:
<img width="512" alt="image" src="https://github.com/user-attachments/assets/00b02222-9498-4228-9277-43cc735da720">
<img width="499" alt="image" src="https://github.com/user-attachments/assets/123b51ba-0879-44e8-9d1a-b37a50ab9e3a">

# Why

- to allow developers to manage their inference providers
- to improve navigation

# Testing done

- tested the new features
- tested the bug fixes
- e2e frontend OK
- unit frontend OK
- backend integration OK

# Decisions made

- Added unique constraint on provider+model+plugin in postrgres
- Added AJV to validate json schema client-side
- Added lodash for easy grouping/sorting in JS
- Skipping two fields on the frontend for now (they don't create errors however because they have a default value in the schema). For future reference, I added a warning in the console to be able to spot missing fields easily:
<img width="299" alt="image" src="https://github.com/user-attachments/assets/aa86852d-6486-47e5-a55f-6802c8cc36a5">


# Checks

- [x] I have tested this code
- [x] I have reviewed my own PR
- [x] I have created an issue for this PR
- [x] I have set a descriptive PR title compliant with [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/)

# Reviewing tips

- Try adding / editing providers, then adding / editing validators

# User facing release notes

- Added the ability to add and manage inference providers from in the settings
- Reworked navigation
- Move account managing to the account select menu
- Fixed other various bugs and issues

denishacquin commented on backend/node/create_nodes/providers_schema.json:
```diff
@@ -191,7 +192,8 @@
-                "const": ""
+                "const": "",
+                "default": ""
               },
```
But kept field hidden in ui however.

AgustinRamiroDiaz commented on backend/database_handler/llm_providers.py:
```diff
@@ -25,10 +27,23 @@ def get_all(self) -> list[LLMProvider]:
-            _to_domain(provider).__dict__
-            for provider in self.session.query(LLMProviderDBModel).all()
-        ]
+        ## TODO: double check this with agustin
```
looks great!

denishacquin reviewed (commented):
self reviewed

cristiam86 commented on backend/protocol_rpc/endpoints.py:
```diff
@@ -127,7 +128,7 @@ def create_validator(
     # TODO: only accept all or none of the config fields
     llm_provider = None
-    if not (config and plugin and plugin_config):
+    if not (config is None and plugin is None and plugin_config is None):
```
@denishacquin this is not what the comment says. This condition is equal to:
if config != None or plugin != None or plugin_config != None

denishacquin replied:
fixed this, thanks for the spot!

cristiam86 reviewed (commented):
I left a small comment and I will approve once I checked&tested the branch.

Would it be possible to add some sort of tests to the new screen?

As a side note, it is kind of confusing that Provider is the name for a provider and a model. Next time let's try to solve this issues before they are not worth it like right now.

denishacquin commented:
> As a side note, it is kind of confusing that Provider is the name for a provider and a model. Next time let's try to solve this issues before they are not worth it like right now.

I agree which is why I was playing around with "**Provider Config**". It is true that the entity we are talking about is a kind of preset so to speak, that combines provider/model/plugin and default config.

Totally open to suggestions, but like you said, it's not dealbreaker right now and since it's an advanced feature, we should be fine for now indeed.

denishacquin commented:
added migration to remove legacy heuristai providers

denishacquin commented:
<img width="717" alt="image" src="https://github.com/user-attachments/assets/c69ff81f-da77-4624-bfcf-836bef55c4d6">
<img width="402" alt="image" src="https://github.com/user-attachments/assets/b764152b-32be-402b-be7f-dc458dc1c8a2">
<img width="826" alt="image" src="https://github.com/user-attachments/assets/723867fb-2de0-4f10-9b09-e2484e47cb80">
<img width="362" alt="image" src="https://github.com/user-attachments/assets/7dc45de4-b327-48af-8083-db48785bceff">

@cristiam86 @AgustinRamiroDiaz made some copy adjustments, added tooltips, added a link to settings from validator modal.

I think this helps with user education, WDYT?

cristiam86 commented on frontend/test/e2e/ManageProviders.spec.ts:
```diff
@@ -0,0 +1,113 @@
+    expect(
+      newProviders.length,
+      'number of providers should be greather than old providers list',
+    ).be.greaterThan(initialProviders.length);
```
@denishacquin we should be more specific and expecting the exact number of providers after the creation

cristiam86 commented on frontend/test/e2e/ManageProviders.spec.ts:
```diff
@@ -0,0 +1,113 @@
+      By.xpath("//button[@data-testid = 'provider-item']"),
+    );
+
+    expect(
```
@denishacquin same as before, please be more specific

cristiam86 reviewed (approved):
Only minor things in e2e tests
